cmd/gateway/s3x/gateway-s3x-ledger-multipart.go

```diff
@@ -0,0 +1,168 @@
+package s3x
+
+import "github.com/ipfs/go-datastore"
+
+/* Design Notes
+---------------
+
+Internal functions should never claim or release locks.
+Any claiming or releasing of locks should be done in the public setter+getter functions.
+The reason for this is so that we can enable easy reuse of internal code.
+*/
+
+/////////////////////
+// SETTER FUNCTINS //
+/////////////////////
+
+// AbortMultipartUpload is used to abort a multipart upload
+func (ls *ledgerStore) AbortMultipartUpload(bucket, multipartID string) error {
+	err := ls.AssertBucketExits(bucket)
+	if err != nil {
+		return err
+	}
+	return ls.DeleteMultipartID(multipartID)
+}
+
+// NewMultipartUpload is used to store the initial start of a multipart upload request
+func (ls *ledgerStore) NewMultipartUpload(multipartID string, info *ObjectInfo) error {
+	bucket := info.GetBucket()
+	defer ls.locker.write(bucket)
```

Comment:
shouldn't this be declared after the `err := ls.assertBucketExists`, since if the bucket doesnt' exist there is no point to trigger a lock write?

Reply:
Actually, it's not needed at all. Since there are no changes to bucket objects.